README
NAME
    Beam::Emitter - Role for event emitting classes

VERSION
    version 1.000

SYNOPSIS
        package My::Emitter;

        use Moo;
        with 'Beam::Emitter';

        sub do_something {
            my ( $self ) = @_;

            # Give event listeners a chance to prevent something
            my $event = $self->emit( "before_something" );
            return if $event->is_default_stopped;

            # ... do something

            # Notify listeners we're done with something
            $self->emit( 'after_something' );
        }

        sub custom_something {
            my ( $self ) = @_;

            # Send arbitrary arguments to our event listener
            $self->emit_args( 'custom_something', "foo", "bar" );
        }

DESCRIPTION
    This role is used by classes that want to emit events to subscribers. A
    subscriber registers interest in an event using the subscribe or on
    methods. Then, the class can emit events to be handled by any listening
    subscribers.

    Using the Beam::Event class, subscribers can stop an event from being
    processed, or prevent the default action from happening.

  Using Beam::Event
    Beam::Event is an event object with some simple methods to allow
    subscribers to influence the handling of the event. By calling the stop
    method, subscribers can stop all futher handling of the event. By
    calling the the stop_default method, subscribers can allow other
    subscribers to be notified about the event, but let the emitter know
    that it shouldn't continue with what it was going to do.

    For example, let's build a door that notifies when someone tries to open
    it. Different instances of a door should allow different checks before
    the door opens, so we'll emit an event before we decide to open.

        package Door;
        use Moo;
        with 'Beam::Emitter';

        sub open {
            my ( $self, $who ) = @_;
            my $event = $self->emit( 'before_open' );
            return if $event->is_default_stopped;
            $self->open_the_door;
        }

        package main;
        my $door = Door->new;
        $door->open;

    Currently, our door will open for anybody. But let's build a door that
    only open opens after noon (to keep us from having to wake up in the
    morning).

        use Time::Piece;
        my $restful_door = Door->new;

        $restful_door->on( before_open => sub {
            my ( $event ) = @_;

            my $time = Time::Piece->now;
            if ( $time->hour < 12 ) {
                $event->stop_default;
            }

        } );

        $restful_door->open;

    By calling stop_default, we set the is_default_stopped flag, which the
    door sees and decides not to open.

  Using Custom Events
    The default "Beam::Event" is really only useful for notifications. If
    you want to give your subscribers some data, you need to create a custom
    event class. This allows you to add attributes and methods to your
    events (with all the type constraints and coersions you want).

    Let's build a door that can keep certain people out. Right now, our door
    doesn't care who is trying to open it, and our subscribers do not get
    enough information to deny entry to certain people.

    So first we need to build an event object that can let our subscribers
    know who is knocking on the door.

        package Door::Knock;
        use Moo;
        extends 'Beam::Event';

        has who => (
            is => 'ro',
            required => 1,
        );

    Now that we can represent who is knocking, let's notify our subscribers.

        package Door;
        use Moo;
        use Door::Knock; # Our emitter must load the class, Beam::Emitter will not
        with 'Beam::Emitter';

        sub open {
            my ( $self, $who ) = @_;
            my $event = $self->emit( 'before_open', class => 'Door::Knock', who => $who );
            return if $event->is_default_stopped;
            $self->open_the_door;
        }

    Finally, let's build a listener that knows who is allowed in the door.

        my $private_door = Door->new;
        $private_door->on( before_open => sub {
            my ( $event ) = @_;

            if ( $event->who ne 'preaction' ) {
                $event->stop_default;
            }

        } );

        $private_door->open;

  Without Beam::Event
    Although checking "is_default_stopped" is completely optional, if you do
    not wish to use the "Beam::Event" object, you can instead call emit_args
    instead of emit to give arbitrary arguments to your listeners.

        package Door;
        use Moo;
        with 'Beam::Emitter';

        sub open {
            my ( $self, $who ) = @_;
            $self->emit_args( 'open', $who );
            $self->open_the_door;
        }

    There's no way to stop the door being opened, but you can at least
    notify someone before it does.

METHODS
  subscribe ( event_name, subref )
    Subscribe to an event from this object. "event_name" is the name of the
    event. "subref" is a subroutine reference that will get either a
    Beam::Event object (if using the emit method) or something else (if
    using the emit_args method).

  on ( event_name, subref )
    Alias for "subscribe".

  unsubscribe ( event_name [, subref ] )
    Unsubscribe from an event. "event_name" is the name of the event.
    "subref" is the single listener subref to be removed. If no subref is
    given, will remove all listeners for this event.

  un ( event_name [, subref ] )
    An alias for "unsubscribe"

  emit ( name, event_args )
    Emit a Beam::Event with the given "name". "event_args" is a list of name
    => value pairs to give to the "Beam::Event" constructor.

    Use the "class" key in "event_args" to specify a different Event class.

  emit_args ( name, callback_args )
    Emit an event with the given "name". "callback_args" is a list that will
    be given directly to each subscribed callback.

    Use this if you want to avoid using Beam::Event, though you miss out on
    the control features like stop and stop default.

SEE ALSO
    Beam::Event
    <http://perladvent.org/2013/2013-12-16.html>
        Coordinating Christmas Dinner with Beam::Emitter by Yanick Champoux.

AUTHOR
    Doug Bell <preaction@cpan.org>

CONTRIBUTOR
    Joel Berger <joel.a.berger@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2015 by Doug Bell.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
Changelog for Beam-Emitter

{{ $NEXT }}

    [Fixed]

    - Passing in an overridden emitter when emitting an event no longer
      uses the emitter's boolean overload. This could cause the wrong
      emitter to be used in the event. Thanks @djerius (Github #25)

1.006     2016-11-23 22:35:48-06:00 America/Chicago

    [Added]

    - Added a couple more cookbook entries to show Beam::Emitter
      patterns

    [Fixed]

    - Events with no listeners now correctly return the Event object
      like they used to. This was a regression starting in v1.004.
      Thanks @eserte and CPAN Testers for the report (Github #24)

1.005     2016-11-22 12:56:41-06:00 America/Chicago

    [Fixed]

    - Custom listener classes are now loaded automatically. This
      prevents a very strange error about the unloaded class not
      inheriting from Beam::Listener, even though it does. Perl doesn't
      know that it does, because the class isn't loaded, and there's no
      good way to detect if a class is loaded or if it failed to load.
      So, to make it easier, Beam::Emitter now simply loads the class
      for you. Thanks @djerius (Github #23)

1.004     2016-11-17 17:28:01-06:00 America/Chicago

    [Added]

    - Allow the event's name to be overridden when emitting the event.
      This is a rare thing to want, but it makes some future cookbook
      items (like a catch-all '*' event) more useful.

    - Allow the event's emitter object to be overridden when emitting
      the event. This, too, is a rare thing to want, but it allows an
      object to more easily proxy the event emitter if needed. Thanks
      @djerius! (Github #19)

    - Added a Cookbook document (Beam::Emitter::Cookbook) to start
      demonstrating some useful patterns that aren't quite useful or
      generally-desired enough to be part of the main distribution
      always. Thanks @yanick for the help! (Github #12)

    - Enhanced how Beam::Emitter tracks listeners by creating
      Beam::Listener objects with the option of adding additional
      metadata to the listener by extending the Beam::Listener class.
      Thanks @djerius! (Github #16)

    - Added a method (`listeners()`) to list the current listeners
      (Beam::Listener objects) for an event. Thanks @djerius! (Github
      #16)

    [Fixed]

    - The Beam::Emitter role no longer pollutes the consumer's namespace
      with things from Types::Standard, Scalar::Util, and Carp. Thanks
      @kentnl! (Github #13)

      This could result in problems if you accidentally depended on this
      behavior. To fix these problems, please import the correct subs
      explicitly.

    - The Beam::Emitter role now allows composing classes to use method
      modifiers for the `subscribe` and `unsubscribe` methods.
      Previously, if the composing class overrode either of those
      methods, those changes would be ignored by the `on` and `un`
      aliases. Now, the `on` and `un` aliases use proper method
      resolution. This means that if you want to modify behaviors of
      your class on subscribe and unsubscribe, you must only modify the
      `subscribe` and `unsubscribe` methods. (Github #18)

1.003     2016-11-11 21:35:53-06:00 America/Chicago

    [Fixed]

    - Fix the unsubscribe callback (returned from the `subscribe`
      method) throwing an exception if the emitter object has already
      been destroyed. Since the unsubscribe callback has a weak
      reference to the emitter object, we have to make sure the
      reference still exists before using it to unsubscribe. Thanks
      @djerius! (Github #14)

    - Fix unsubscribing from an event during an event not firing all the
      rest of the events properly. When using the unsubscribe callback
      to create a one-time event handler, the emitter could miss the
      next event listener (since we were looping over the listeners
      array and modifying it in the loop). Now we loop over a copy of
      the listeners array so that the original can be modified to remove
      the listener we want. Thanks @djerius! (Github #15)

1.002     2016-09-02 18:58:23-05:00 America/Chicago

 - Update distribution settings

1.001 2016-07-19T16:51:49
 - use a better meta-syntactic variable
 - fix circular references in unsubscribe callback
 - return subref that removes the current subscriber
 - upgrade perl in travis config
 - remove ModuleBuild to prevent toolchain confusion

1.000 2015-05-08T04:34:31Z
 - fix typo in documentation
 - remove private attribute from public documentation

0.009 2015-05-08T04:34:01Z
 - fix typo in documentation
 - remove private attribute from public documentation

0.008 2015-03-02T02:33:00Z
 - add note about loading custom event classes
 - add dzil plugin for prereqs and compile tests
 - remove root dotfiles and root ini files from dist
 - switch to dzil Git::GatherDir
 - fix contributors dependency

0.007 2015-01-04T14:30:47Z
 - fix docs to remove return value from emit_args()
 - fix link text to read more like a paragraph
 - add tagline for advent article

0.006 2015-01-04T02:55:56Z
 - release v0.006
 - fix POD errors and add SEE ALSO sections

0.005 2015-01-04T02:46:14Z
 - release v0.005
 - add tutorial-style documentation
 - add link to #beam on irc.perl.org

0.004 2015-01-04T01:55:02Z
 - release v0.004
 - update Types::Standard to 0.008
 - update copyright year
 - add contributors dzil plugin for proper attribution
 - only ignore bundles in the dist root
 - only remove .ini files from the dist root
 - add dzil tag annotation with name, date, and changelog

0.003 2014-11-16T01:42:26Z
 - release v0.003
 - allow cpanfile to be dirty
 - add new build artifacts
 - fix dist.ini after copying from Beam-Wire
 - interleave and update documentation for weaving
 - clarify which method uses Beam::Event
 - add more missing dependencies
 - switch to Type::Tiny
 - add missing Test::More dependency
 - update dist.ini and add travis-ci
 - add emit_args to give args directly to listeners

t/CustomListener.pm
package t::CustomListener;

use Moo;
extends 'Beam::Listener';

has attr => ( is => 'ro', required => 1 );

1;



Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.023.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "Role for event emitting classes",
  "dist_author" => [
    "Doug Bell <preaction\@cpan.org>"
  ],
  "dist_name" => "Beam-Emitter",
  "dist_version" => "1.003",
  "license" => "perl",
  "module_name" => "Beam::Emitter",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Moo" => 0,
    "Scalar::Util" => 0,
    "Types::Standard" => "0.008",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::Exception" => 0,
    "Test::More" => 0
  }
);


my %fallback_build_requires = (
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Build" => "0.28",
  "Test::Exception" => 0,
  "Test::More" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Beam/Event.pm
package Beam::Event;
our $VERSION = '1.008';
# ABSTRACT: Base Event class

=head1 SYNOPSIS

    # My::Emitter consumes the Beam::Emitter role
    my $emitter = My::Emitter->new;
    $emitter->on( "foo", sub {
        my ( $event ) = @_;
        print "Foo happened!\n";
        # stop this event from continuing
        $event->stop;
    } );
    my $event = $emitter->emit( "foo" );

=head1 DESCRIPTION

This is the base event class for C<Beam::Emitter> objects.

The base class is only really useful for notifications. Create a subclass
to add data attributes.

=head1 SEE ALSO

=over 4

=item L<Beam::Emitter>

=back

=cut

use strict;
use warnings;

use Moo;
use Types::Standard qw(:all);

=attr name

The name of the event. This is the string that is given to L<Beam::Emitter/on>.

=cut

has name => (
    is          => 'ro',
    isa         => Str,
    required    => 1,
);

=attr emitter

The emitter of this event. This is the object that created the event.

=cut

has emitter => (
    is          => 'ro',
    isa         => ConsumerOf['Beam::Emitter'],
    required    => 1,
);

=attr is_default_stopped

This is true if anyone called L</stop_default> on this event.

Your L<emitter|Beam::Emitter> should check this attribute before trying to do
what the event was notifying about.

=cut

has is_default_stopped => (
    is          => 'rw',
    isa         => Bool,
    default     => sub { 0 },
);

=attr is_stopped

This is true if anyone called L</stop> on this event.

When using L<the emit method|Beam::Emitter/emit>, this is checked automatically
after every callback, and event processing is stopped if this is true.

=cut

has is_stopped => (
    is          => 'rw',
    isa         => Bool,
    default     => sub { 0 },
);

=method stop_default ()

Calling this will cause the default behavior of this event to be stopped.

B<NOTE:> Your event-emitting object must check L</is_default_stopped> for this
behavior to work.

=cut

sub stop_default {
    my ( $self ) = @_;
    $self->is_default_stopped( 1 );
}

=method stop ()

Calling this will immediately stop any further processing of this event.
Also calls L</stop_default>.

=cut

sub stop {
    my ( $self ) = @_;
    $self->stop_default;
    $self->is_stopped( 1 );
}

1;
__END__



lib/Beam/Emitter.pm
package Beam::Emitter;
our $VERSION = '1.008';
# ABSTRACT: Role for event emitting classes

=head1 SYNOPSIS

    # A simple custom event class to perform data validation
    { package My::Event;
        use Moo;
        extends 'Beam::Event';
        has data => ( is => 'ro' );
    }

    # A class that reads and writes data, allowing event handlers to
    # process the data
    { package My::Emitter;
        use Moo;
        with 'Beam::Emitter';

        sub write_data {
            my ( $self, @data ) = @_;

            # Give event listeners a chance to perform further processing of
            # data
            my $event = $self->emit( "process_data",
                class => 'My::Event',
                data => \@data,
            );

            # Give event listeners a chance to stop the write
            return if $event->is_default_stopped;

            # Write the data
            open my $file, '>', 'output';
            print { $file } @data;
            close $file;

            # Notify listeners we're done writing and send them the data
            # we wrote
            $self->emit( 'after_write', class => 'My::Event', data => \@data );
        }
    }

    # An event handler that increments every input value in our data
    sub increment {
        my ( $event ) = @_;
        my $data = $event->data;
        $_++ for @$data;
    }

    # An event handler that performs data validation and stops the
    # processing if invalid
    sub prevent_negative {
        my ( $event ) = @_;
        my $data = $event->data;
        $event->prevent_default if grep { $_ < 0 } @$data;
    }

    # An event handler that logs the data to STDERR after we've written in
    sub log_data {
        my ( $event ) = @_;
        my $data = $event->data;
        print STDERR "Wrote data: " . join( ',', @$data );
    }

    # Wire up our event handlers to a new processing object
    my $processor = My::Emitter->new;
    $processor->on( process_data => \&increment );
    $processor->on( process_data => \&prevent_negative );
    $processor->on( after_write => \&log_data );

    # Process some data
    $processor->process_data( 1, 2, 3, 4, 5 );
    $processor->process_data( 1, 3, 7, -9, 11 );

    # Log data before and after writing
    my $processor = My::Emitter->new;
    $processor->on( process_data => \&log_data );
    $processor->on( after_write => \&log_data );

=head1 DESCRIPTION

This role is used by classes that want to add callback hooks to allow
users to add new behaviors to their objects. These hooks are called
"events". A subscriber registers a callback for an event using the
L</subscribe> or L</on> methods. Then, the class can call those
callbacks by L<emitting an event with the emit() method|/emit>.

Using the L<Beam::Event> class, subscribers can stop an event from being
processed, or prevent the default action from happening.

=head2 Using Beam::Event

L<Beam::Event> is an event object with some simple methods to allow subscribers
to influence the handling of the event. By calling L<the stop
method|Beam::Event/stop>, subscribers can stop all futher handling of the
event. By calling the L<the stop_default method|Beam::Event/stop_default>,
subscribers can allow other subscribers to be notified about the event, but let
the emitter know that it shouldn't continue with what it was going to do.

For example, let's build a door that notifies when someone tries to open it.
Different instances of a door should allow different checks before the door
opens, so we'll emit an event before we decide to open.

    package Door;
    use Moo;
    with 'Beam::Emitter';

    sub open {
        my ( $self, $who ) = @_;
        my $event = $self->emit( 'before_open' );
        return if $event->is_default_stopped;
        $self->open_the_door;
    }

    package main;
    my $door = Door->new;
    $door->open;

Currently, our door will open for anybody. But let's build a door that only
open opens after noon (to keep us from having to wake up in the morning).

    use Time::Piece;
    my $restful_door = Door->new;

    $restful_door->on( before_open => sub {
        my ( $event ) = @_;

        my $time = Time::Piece->now;
        if ( $time->hour < 12 ) {
            $event->stop_default;
        }

    } );

    $restful_door->open;

By calling L<stop_default|Beam::Event/stop_default>, we set the
L<is_default_stopped|Beam::Event/is_default_stopped> flag, which the door sees
and decides not to open.

=head2 Using Custom Events

The default C<Beam::Event> is really only useful for notifications. If you want
to give your subscribers some data, you need to create a custom event class.
This allows you to add attributes and methods to your events (with all
the type constraints and coersions you want).

Let's build a door that can keep certain people out. Right now, our door
doesn't care who is trying to open it, and our subscribers do not get enough
information to deny entry to certain people.

So first we need to build an event object that can let our subscribers know
who is knocking on the door.

    package Door::Knock;
    use Moo;
    extends 'Beam::Event';

    has who => (
        is => 'ro',
        required => 1,
    );

Now that we can represent who is knocking, let's notify our subscribers.

    package Door;
    use Moo;
    use Door::Knock; # Our emitter must load the class, Beam::Emitter will not
    with 'Beam::Emitter';

    sub open {
        my ( $self, $who ) = @_;
        my $event = $self->emit( 'before_open', class => 'Door::Knock', who => $who );
        return if $event->is_default_stopped;
        $self->open_the_door;
    }

Finally, let's build a listener that knows who is allowed in the door.

    my $private_door = Door->new;
    $private_door->on( before_open => sub {
        my ( $event ) = @_;

        if ( $event->who ne 'preaction' ) {
            $event->stop_default;
        }

    } );

    $private_door->open;

=head2 Without Beam::Event

Although checking C<is_default_stopped> is completely optional, if you do not
wish to use the C<Beam::Event> object, you can instead call L<emit_args>
instead of L<emit> to give arbitrary arguments to your listeners.

    package Door;
    use Moo;
    with 'Beam::Emitter';

    sub open {
        my ( $self, $who ) = @_;
        $self->emit_args( 'open', $who );
        $self->open_the_door;
    }

There's no way to stop the door being opened, but you can at least notify
someone before it does.

=head1 SEE ALSO

=over 4

=item L<Beam::Event>

=item L<Beam::Emitter::Cookbook>

This document contains some useful patterns for your event emitters and
listeners.

=item L<http://perladvent.org/2013/2013-12-16.html>

Coordinating Christmas Dinner with Beam::Emitter by Yanick Champoux.

=back

=cut

use strict;
use warnings;

use Types::Standard qw(:all);
use Scalar::Util qw( weaken refaddr );
use Carp qw( croak );
use Beam::Event;
use Module::Runtime qw( use_module );
use Moo::Role; # Put this last to ensure proper, automatic cleanup


# The event listeners on this object, a hashref of arrayrefs of
# EVENT_NAME => [ Beam::Listener object, ... ]

has _listeners => (
    is      => 'ro',
    isa     => HashRef,
    default => sub { {} },
);

=method subscribe ( event_name, subref, [ %args ] )

Subscribe to an event from this object. C<event_name> is the name of the event.
C<subref> is a subroutine reference that will get either a L<Beam::Event> object
(if using the L<emit> method) or something else (if using the L<emit_args> method).

Returns a coderef that, when called, unsubscribes the new subscriber.

    my $unsubscribe = $emitter->subscribe( open_door => sub {
        warn "ding!";
    } );
    $emitter->emit( 'open_door' );  # ding!
    $unsubscribe->();
    $emitter->emit( 'open_door' );  # no ding

This unsubscribe subref makes it easier to stop our subscription in a safe,
non-leaking way:

    my $unsub;
    $unsub = $emitter->subscribe( open_door => sub {
        $unsub->(); # Only handle one event
    } );
    $emitter->emit( 'open_door' );

The above code does not leak memory, but the following code does:

    # Create a memory cycle which must be broken manually
    my $cb;
    $cb = sub {
        my ( $event ) = @_;
        $event->emitter->unsubscribe( open_door => $cb ); # Only handle one event
        # Because the callback sub ($cb) closes over a reference to itself
        # ($cb), it can never be cleaned up unless something breaks the
        # cycle explicitly.
    };
    $emitter->subscribe( open_door => $cb );
    $emitter->emit( 'open_door' );

The way to fix this second example is to explicitly C<undef $cb> inside the callback
sub. Forgetting to do that will result in a leak. The returned unsubscribe coderef
does not have this issue.

By default, the emitter only stores the subroutine reference in an
object of class L<Beam::Listener>.  If more information should be
stored, create a custom subclass of L<Beam::Listener> and use C<%args>
to specify the class name and any attributes to be passed to its
constructor:

  {
    package MyListener;
    extends 'Beam::Listener';

    # add metadata with subscription time
    has sub_time => is ( 'ro',
			  init_arg => undef,
			  default => sub { time() },
    );
  }

  # My::Emitter consumes the Beam::Emitter role
  my $emitter = My::Emitter->new;
  $emitter->on( "foo",
    sub { print "Foo happened!\n"; },
   class => MyListener
  );

The L</listeners> method can be used to examine the subscribed listeners.


=cut

sub subscribe {
    my ( $self, $name, $sub, %args ) = @_;

    my $class = delete $args{ class } || "Beam::Listener";
    croak( "listener object must descend from Beam::Listener" )
      unless use_module($class)->isa( 'Beam::Listener' );

    my $listener = $class->new( %args, callback => $sub );

    push @{ $self->_listeners->{$name} }, $listener;
    weaken $self;
    weaken $sub;
    return sub {
        $self->unsubscribe($name => $sub)
	  if defined $self;
    };
}

=method on ( event_name, subref )

An alias for L</subscribe>. B<NOTE>: Do not use this alias for method
modifiers! If you want to override behavior, override C<subscribe>.

=cut

sub on { shift->subscribe( @_ ) }

=method unsubscribe ( event_name [, subref ] )

Unsubscribe from an event. C<event_name> is the name of the event. C<subref> is
the single listener subref to be removed. If no subref is given, will remove
all listeners for this event.

=cut

sub unsubscribe {
    my ( $self, $name, $sub ) = @_;
    if ( !$sub ) {
        delete $self->_listeners->{$name};
    }
    else {
        my $listeners = $self->_listeners->{$name};
        my $idx = 0;
        $idx++ until $idx > $#{$listeners} or refaddr $listeners->[$idx]->callback eq refaddr $sub;
        if ( $idx > $#{$listeners} ) {
            croak "Could not find sub in listeners";
        }
        splice @{$self->_listeners->{$name}}, $idx, 1;
    }
    return;
}

=method un ( event_name [, subref ] )

An alias for L</unsubscribe>. B<NOTE>: Do not use this alias for method
modifiers! If you want to override behavior, override C<unsubscribe>.

=cut

sub un { shift->unsubscribe( @_ ) }

=method emit ( name, event_args )

Emit a L<Beam::Event> with the given C<name>. C<event_args> is a list of name => value
pairs to give to the C<Beam::Event> constructor.

Use the C<class> key in C<event_args> to specify a different Event class.

=cut

sub emit {
    my ( $self, $name, %args ) = @_;

    my $class = delete $args{ class } || "Beam::Event";
    $args{ emitter  } = $self if ! defined $args{ emitter };
    $args{ name     } ||= $name;
    my $event = $class->new( %args );

    return $event unless exists $self->_listeners->{$name};

    # don't use $self->_listeners->{$name} directly, as callbacks may unsubscribe
    # from $name, changing the array, and confusing the for loop
    my @listeners = @{ $self->_listeners->{$name} };

    for my $listener ( @listeners  ) {
        $listener->callback->( $event );
        last if $event->is_stopped;
    }
    return $event;
}

=method emit_args ( name, callback_args )

Emit an event with the given C<name>. C<callback_args> is a list that will be given
directly to each subscribed callback.

Use this if you want to avoid using L<Beam::Event>, though you miss out on the control
features like L<stop|Beam::Event/stop> and L<stop default|Beam::Event/stop_default>.

=cut

sub emit_args {
    my ( $self, $name, @args ) = @_;

    return unless exists $self->_listeners->{$name};

    # don't use $self->_listeners->{$name} directly, as callbacks may unsubscribe
    # from $name, changing the array, and confusing the for loop
    my @listeners = @{ $self->_listeners->{$name} };

    for my $listener ( @listeners ) {
        $listener->callback->( @args );
    }
    return;
}

=method listeners ( event_name )

Returns a list containing the listeners which have subscribed to the
specified event from this emitter.  The list elements are either
instances of L<Beam::Listener> or of custom classes specified in calls
to L</subscribe>.

=cut

sub listeners {

    my ( $self, $name ) = @_;

    return @{ $self->_listeners->{$name} || [] };
}

1;
__END__



lib/Beam/Listener.pm
package Beam::Listener;
our $VERSION = '1.008';

=head1 SYNOPSIS

  package MyListener;

  extends 'Beam::Listener';


  # add metadata with subscription time
  has sub_time => is ( 'ro',
                        init_arg => undef,
                        default => sub { time() },
  );

   # My::Emitter consumes the Beam::Emitter role
   my $emitter = My::Emitter->new;
   $emitter->on( "foo", sub {
        my ( $event ) = @_;
        print "Foo happened!\n";
        # stop this event from continuing
        $event->stop;
    },
    class => MyListener
    );


=head1 DESCRIPTION

This is the base class used by C<Beam::Emitter> objects to store information
about listeners. Create a subclass to add data attributes.

=head1 SEE ALSO

=over 4

=item L<Beam::Emitter>

=back

=cut

use strict;
use warnings;

use Types::Standard qw(:all);
use Moo;

=attr code

A coderef which will be invoked when the event is distributed.

=cut

has callback => (
    is  => 'ro',
    isa => CodeRef,
    required => 1,
);

1;

__END__



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.008.
use strict;
use warnings;



use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Role for event emitting classes",
  "AUTHOR" => "Doug Bell <preaction\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Beam-Emitter",
  "LICENSE" => "perl",
  "NAME" => "Beam::Emitter",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Module::Runtime" => 0,
    "Moo" => 0,
    "Scalar::Util" => 0,
    "Types::Standard" => "0.008",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::API" => 0,
    "Test::Fatal" => 0,
    "Test::Lib" => 0,
    "Test::More" => 0
  },
  "VERSION" => "1.008",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Runtime" => 0,
  "Moo" => 0,
  "Scalar::Util" => 0,
  "Test::API" => 0,
  "Test::Fatal" => 0,
  "Test::Lib" => 0,
  "Test::More" => 0,
  "Types::Standard" => "0.008",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
